cmd/promtool/analyze.go

```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		return stats, err
+	}
+
+	sumBucketsChanged := 0
+	for timeIdx := 1; timeIdx < numSamples; timeIdx++ {
+		curr, err := getBucketCountsAtTime(matrix, numBuckets, timeIdx)
+		if err != nil {
+			return stats, err
+		}
+
+		countBucketsChanged := 0
+		for bIdx := range matrix {
+			if curr[bIdx] > prev[bIdx] {
+				countBucketsChanged++
+			} else if curr[bIdx] < prev[bIdx] {
+				countBucketsChanged++
+				fmt.Printf("unexpected decrease in bucket %d at time %d\n", bIdx, timeIdx)
+				// return stats, fmt.Errorf("unexpected decrease in bucket %d at time %d", bIdx, timeIdx)
+			}
+		}
+
+		histo.update(countBucketsChanged)
+
+		sumBucketsChanged += countBucketsChanged
+		if stats.min > countBucketsChanged {
+			stats.min = countBucketsChanged
+		}
+		if stats.max < countBucketsChanged {
+			stats.max = countBucketsChanged
+		}
+
+		prev = curr
+	}
+	stats.avg = float64(sumBucketsChanged) / float64(numSamples-1)
+	return stats, nil
+}
+
+func sortMatrix(matrix model.Matrix) {
+	sort.SliceStable(matrix, func(i, j int) bool {
+		return getLe(matrix[i]) < getLe(matrix[j])
+	})
+}
+
+func getLe(vector *model.SampleStream) float64 {
+	lbs := model.LabelSet(vector.Metric)
+	le, _ := strconv.ParseFloat(string(lbs["le"]), 64)
+	return le
+}
+
+func getBucketCountsAtTime(matrix model.Matrix, numBuckets, timeIdx int) ([]int, error) {
+	counts := make([]int, numBuckets)
+	counts[0] = int(matrix[0].Values[timeIdx].Value)
+	for i, bucket := range matrix[1:] {
+		curr := bucket.Values[timeIdx]
+		prev := matrix[i].Values[timeIdx]
+		// Assume the results are nicely aligned.
+		if curr.Timestamp != prev.Timestamp {
+			return counts, fmt.Errorf("matrix result is not time aligned")
+		}
+		counts[i+1] = int(curr.Value - prev.Value)
+	}
+	return counts, nil
+}
+
+type bucketBounds struct {
+	boundaries   int32
+	upper, lower float64
+}
+
+func makeBucketBounds(b *model.HistogramBucket) bucketBounds {
+	return bucketBounds{
+		boundaries: b.Boundaries,
+		upper:      float64(b.Upper),
+		lower:      float64(b.Lower),
+	}
+}
+
+func calcNativeBucketStatistics(matrix model.Matrix, step time.Duration, histo *statshistogram) (*statistics, error) {
+	stats := &statistics{
+		min:            math.MaxInt,
+		scrapeInterval: step,
+	}
+
+	overall := make(map[bucketBounds]struct{})
+	sumBucketsChanged := 0
+	for _, vector := range matrix {
+		prev := make(map[bucketBounds]float64)
+		for _, bucket := range vector.Histograms[0].Histogram.Buckets {
+			bb := makeBucketBounds(bucket)
+			prev[bb] = float64(bucket.Count)
+			overall[bb] = struct{}{}
+		}
+		for _, histogram := range vector.Histograms[1:] {
+			curr := make(map[bucketBounds]float64)
+			for _, bucket := range histogram.Histogram.Buckets {
+				bb := makeBucketBounds(bucket)
+				curr[bb] = float64(bucket.Count)
+				overall[bb] = struct{}{}
+			}
+			countBucketsChanged := 0
+			for bucket, currCount := range curr {
+				prevCount, ok := prev[bucket]
+				if !ok {
+					countBucketsChanged++
```

Comment:
Again, I don't think that counting changed buckets is a good indicator for resource cost.

Unpopulated buckets (which have a very small cost, ideally zero) are already not returned by the API. Just counting returned buckets should be a better indicator. (But note that a native histogram bucket is less costly than a classic histogram bucket. The former is just an additional piece of data in the complex sample structure while the latter is its own series.)

If you wanted to, you could try to guess if the bucket is encoded with XOR or with double-delta, and then estimate storage cost, but that gets really close into reimplementing the whole encoding of histograms. As already discussed separately, to go down that alley it would be better to analyze the histograms on the block level, where you can directly see how much storage they take.